File: server/src/routes/admin/spaces.ts
/**
 * Admin Space/Synapse Inspection Routes
 *
 * Browse, search, and inspect synapses.
 */

import { Router, Request, Response } from 'express'
import { db } from '../../db/index.js'
import { asyncHandler, sendError } from '../../middleware/errorHandler.js'

const router = Router()

interface SpaceRow {
  id: string
  state: string
  synapse_type: string
  position_x: number
  position_y: number
  position_z: number
  points_required: number
  points_accumulated: number
  agi_reward: number
  discovered_at: number | null
}

/**
 * GET /api/admin/spaces
 * List spaces with pagination, search, and filters
 */
router.get('/', asyncHandler(async (req: Request, res: Response) => {
  const page = Math.max(1, parseInt(req.query.page as string) || 1)
  const limit = Math.min(100, Math.max(1, parseInt(req.query.limit as string) || 50))
  const offset = (page - 1) * limit
  const search = (req.query.search as string) || ''
  const state = req.query.state as string // 'undiscovered', 'being_solved', 'discovered'
  const synapseType = req.query.synapseType as string
  const sortBy = (req.query.sortBy as string) || 'created_at'
  const sortOrder = (req.query.sortOrder as string)?.toUpperCase() === 'ASC' ? 'ASC' : 'DESC'

  const conditions: string[] = []
  const params: (string | number)[] = []

  if (search) {
    conditions.push('id LIKE ?')
    params.push(`${search}%`)
  }

  if (state) {
    conditions.push('state = ?')
    params.push(state)
  }

  if (synapseType) {
    conditions.push('synapse_type = ?')
    params.push(synapseType)
  }

  const whereClause = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : ''

  const validSortColumns = ['discovered_at', 'points_accumulated', 'agi_reward', 'synapse_type', 'state']
  const sortColumn = validSortColumns.includes(sortBy) ? sortBy : 'discovered_at'

  const countResult = db.prepare(`SELECT COUNT(*) as total FROM spaces ${whereClause}`).get(...params) as { total: number }

  const spaces = db.prepare(`
    SELECT id, state, synapse_type, position_x, position_y, position_z,
           points_required, points_accumulated, agi_reward, discovered_at
    FROM spaces
    ${whereClause}
    ORDER BY ${sortColumn} ${sortOrder}
    LIMIT ? OFFSET ?
  `).all(...params, limit, offset) as SpaceRow[]

  res.json({
    spaces: spaces.map(s => ({
      id: s.id,
      state: s.state,
      synapseType: s.synapse_type,
      position: { x: s.position_x, y: s.position_y, z: s.position_z },
      pointsRequired: s.points_required,
      pointsAccumulated: s.points_accumulated,
      progress: s.points_required > 0 ? (s.points_accumulated / s.points_required * 100).toFixed(2) : '0',
      agiReward: s.agi_reward,
      discoveredAt: s.discovered_at,
    })),
    pagination: {
      page,
      limit,
      total: countResult.total,
      totalPages: Math.ceil(countResult.total / limit),
    },
  })
}))

/**
 * GET /api/admin/spaces/status/in-progress
 * Get spaces currently being explored
 * NOTE: Must be before /:id route to avoid conflict
 */
router.get('/status/in-progress', asyncHandler(async (req: Request, res: Response) => {
  const limit = Math.min(100, Math.max(10, parseInt(req.query.limit as string) || 50))

  const spaces = db.prepare(`
    SELECT s.id, s.synapse_type, s.points_required, s.points_accumulated, s.agi_reward,
           COUNT(se.user_id) as explorer_count
    FROM spaces s
    LEFT JOIN synapse_explorers se ON s.id = se.synapse_id
    WHERE s.state = 'being_solved'
    GROUP BY s.id
    ORDER BY s.points_accumulated DESC
    LIMIT ?
  `).all(limit) as Array<{
    id: string
    synapse_type: string
    points_required: number
    points_accumulated: number
    agi_reward: number
    explorer_count: number
  }>

  res.json({
    spaces: spaces.map(s => ({
      id: s.id,
      synapseType: s.synapse_type,
      pointsRequired: s.points_required,
      pointsAccumulated: s.points_accumulated,
      progress: s.points_required > 0 ? (s.points_accumulated / s.points_required * 100).toFixed(2) : '0',
      agiReward: s.agi_reward,
      explorerCount: s.explorer_count,
    })),
  })
}))

/**
 * GET /api/admin/spaces/:id
 * Get detailed space info including explorers
 */
router.get('/:id', asyncHandler(async (req: Request, res: Response) => {
  const { id } = req.params

  const space = db.prepare(`
    SELECT * FROM spaces WHERE id = ?
  `).get(id) as SpaceRow | undefined

  if (!space) {
    sendError(res, 404, 'Space not found')
    return
  }

  // Get current explorers
  const explorers = db.prepare(`
    SELECT se.user_id, se.points_contributed, se.joined_at, u.wallet
    FROM synapse_explorers se
    JOIN users u ON se.user_id = u.id
    WHERE se.synapse_id = ?
    ORDER BY se.points_contributed DESC
  `).all(id) as Array<{
    user_id: string
    points_contributed: number
    joined_at: number
    wallet: string
  }>

  // Get ships at this synapse
  const ships = db.prepare(`
    SELECT id, name, owner_id, state
    FROM agents
    WHERE current_space_id = ? OR target_space_id = ?
  `).all(id, id) as Array<{
    id: string
    name: string
    owner_id: string
    state: string
  }>

  res.json({
    space: {
      id: space.id,
      state: space.state,
      synapseType: space.synapse_type,
      position: { x: space.position_x, y: space.position_y, z: space.position_z },
      pointsRequired: space.points_required,
      pointsAccumulated: space.points_accumulated,
      progress: space.points_required > 0 ? (space.points_accumulated / space.points_required * 100).toFixed(2) : '0',
      agiReward: space.agi_reward,
      discoveredAt: space.discovered_at,
    },
    explorers: explorers.map(e => ({
      userId: e.user_id,
      wallet: e.wallet,
      pointsContributed: e.points_contributed,
      joinedAt: e.joined_at,
    })),
    ships: ships.map(s => ({
      id: s.id,
      name: s.name,
      ownerId: s.owner_id,
      state: s.state,
    })),
  })
}))

export default router

File: server-go/internal/handlers/admin/spaces.go
package admin

import (
	"errors"
	"strconv"

	"teneo/server-go/internal/database"
	"teneo/server-go/internal/database/generated"

	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5"
)

// ListSpaces returns paginated list of spaces/synapses
func ListSpaces(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	page, _ := strconv.Atoi(c.Query("page", "1"))
	limit, _ := strconv.Atoi(c.Query("limit", "50"))
	state := c.Query("state", "")
	synapseType := c.Query("type", "")
	if synapseType == "" {
		synapseType = c.Query("synapseType", "")
	}
	offset := (page - 1) * limit

	var spaces []generated.Space
	var total int64

	if state != "" || synapseType != "" {
		// Use raw SQL for dynamic filtering since sqlc doesn't support dynamic WHERE
		countQuery := "SELECT COUNT(*) FROM spaces WHERE 1=1"
		dataQuery := `SELECT id, position_x, position_y, position_z, region, zone, synapse_count,
		                     state, discovered_at, synapse_type, points_required, points_accumulated,
		                     current_eta_minutes, sector_id, agi_reward, brain_xp_reward
		              FROM spaces WHERE 1=1`
		var args []interface{}
		argIdx := 1

		if state != "" {
			countQuery += " AND state = $" + strconv.Itoa(argIdx)
			dataQuery += " AND state = $" + strconv.Itoa(argIdx)
			args = append(args, state)
			argIdx++
		}
		if synapseType != "" {
			countQuery += " AND synapse_type = $" + strconv.Itoa(argIdx)
			dataQuery += " AND synapse_type = $" + strconv.Itoa(argIdx)
			args = append(args, synapseType)
			argIdx++
		}

		// Count
		countRow := store.Pool.QueryRow(ctx, countQuery, args...)
		if err := countRow.Scan(&total); err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to count spaces"})
		}

		// Data with pagination
		dataQuery += " ORDER BY id LIMIT $" + strconv.Itoa(argIdx) + " OFFSET $" + strconv.Itoa(argIdx+1)
		args = append(args, limit, offset)

		rows, err := store.Pool.Query(ctx, dataQuery, args...)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to query spaces"})
		}
		defer rows.Close()

		for rows.Next() {
			var s generated.Space
			if err := rows.Scan(
				&s.ID, &s.PositionX, &s.PositionY, &s.PositionZ, &s.Region, &s.Zone,
				&s.SynapseCount, &s.State, &s.DiscoveredAt, &s.SynapseType,
				&s.PointsRequired, &s.PointsAccumulated, &s.CurrentEtaMinutes,
				&s.SectorID, &s.AgiReward, &s.BrainXpReward,
			); err != nil {
				return c.Status(500).JSON(fiber.Map{"error": "Failed to scan space"})
			}
			spaces = append(spaces, s)
		}
		if err := rows.Err(); err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to iterate spaces"})
		}
	} else {
		var err error
		total, err = store.Queries.GetSpaceCount(ctx)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to count spaces"})
		}

		spaces, err = store.Queries.ListSpaces(ctx, generated.ListSpacesParams{
			Limit:  int32(limit),
			Offset: int32(offset),
		})
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to list spaces"})
		}
	}

	result := make([]fiber.Map, len(spaces))
	for i, space := range spaces {
		discoveredAt := int64(0)
		if space.DiscoveredAt != nil {
			discoveredAt = *space.DiscoveredAt
		}
		progress := float64(0)
		if space.PointsRequired > 0 {
			progress = float64(space.PointsAccumulated) / float64(space.PointsRequired) * 100
		}

		result[i] = fiber.Map{
			"id":                space.ID,
			"synapseType":       space.SynapseType,
			"state":             space.State,
			"position":          fiber.Map{"x": space.PositionX, "y": space.PositionY, "z": space.PositionZ},
			"pointsRequired":    space.PointsRequired,
			"pointsAccumulated": space.PointsAccumulated,
			"progress":          progress,
			"agiReward":         space.AgiReward,
			"discoveredAt":      discoveredAt,
		}
	}

	totalPages := (total + int64(limit) - 1) / int64(limit)
	return c.JSON(fiber.Map{
		"spaces": result,
		"pagination": fiber.Map{
			"page":       page,
			"limit":      limit,
			"total":      total,
			"totalPages": totalPages,
		},
	})
}

// GetInProgressSpaces returns spaces currently being explored
func GetInProgressSpaces(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	spaces, err := store.Queries.GetInProgressSpaces(ctx, generated.GetInProgressSpacesParams{
		Limit:  1000,
		Offset: 0,
	})
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get in-progress spaces"})
	}

	result := make([]fiber.Map, len(spaces))
	for i, space := range spaces {
		// Count explorers for this space
		explorerCount, err := store.Queries.GetSynapseExplorerCount(ctx, space.ID)
		if err != nil {
			explorerCount = 0
		}

		progress := float64(0)
		if space.PointsRequired > 0 {
			progress = float64(space.PointsAccumulated) / float64(space.PointsRequired) * 100
		}

		result[i] = fiber.Map{
			"id":                space.ID,
			"synapseType":       space.SynapseType,
			"state":             space.State,
			"totalPoints":       space.PointsRequired,
			"pointsAccumulated": space.PointsAccumulated,
			"progress":          progress,
			"explorerCount":     explorerCount,
		}
	}

	return c.JSON(fiber.Map{"inProgress": result, "count": len(result)})
}

// GetSpaceDetail returns detailed info about a specific space
func GetSpaceDetail(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()
	spaceID := c.Params("id")

	space, err := store.Queries.GetSpace(ctx, spaceID)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return c.Status(404).JSON(fiber.Map{"error": "Space not found"})
		}
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get space"})
	}

	// Get current explorers (only for active synapses)
	var explorers []fiber.Map
	if space.State == "being_solved" {
		// Use raw SQL for the join query
		rows, err := store.Pool.Query(ctx,
			`SELECT se.id, se.ship_id, se.user_id, se.points_per_minute, se.points_contributed, se.joined_at,
			        a.name, u.wallet
			 FROM synapse_explorers se
			 LEFT JOIN agents a ON se.ship_id = a.id
			 LEFT JOIN users u ON se.user_id = u.id
			 WHERE se.synapse_id = $1`, spaceID)
		if err == nil {
			defer rows.Close()
			for rows.Next() {
				var (
					id                string
					shipID            string
					userID            string
					pointsPerMinute   int32
					pointsContributed int32
					joinedAt          int64
					shipName          *string
					wallet            *string
				)
				if err := rows.Scan(&id, &shipID, &userID, &pointsPerMinute, &pointsContributed, &joinedAt, &shipName, &wallet); err != nil {
					continue
				}
				sn := ""
				if shipName != nil {
					sn = *shipName
				}
				w := ""
				if wallet != nil {
					w = *wallet
				}
				explorers = append(explorers, fiber.Map{
					"id":                id,
					"shipId":            shipID,
					"shipName":          sn,
					"userId":            userID,
					"userWallet":        w,
					"pointsPerMin":      pointsPerMinute,
					"pointsContributed": pointsContributed,
					"joinedAt":          joinedAt,
				})
			}
		}
	}

	progress := float64(0)
	if space.PointsRequired > 0 {
		progress = float64(space.PointsAccumulated) / float64(space.PointsRequired) * 100
	}
	discoveredAt := int64(0)
	if space.DiscoveredAt != nil {
		discoveredAt = *space.DiscoveredAt
	}

	return c.JSON(fiber.Map{
		"id":                space.ID,
		"synapseType":       space.SynapseType,
		"state":             space.State,
		"positionX":         space.PositionX,
		"positionY":         space.PositionY,
		"positionZ":         space.PositionZ,
		"pointsRequired":    space.PointsRequired,
		"pointsAccumulated": space.PointsAccumulated,
		"progress":          progress,
		"discoveredAt":      discoveredAt,
		"currentExplorers":  explorers,
	})
}
